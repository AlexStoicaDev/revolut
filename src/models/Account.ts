import { Currency } from './Currency';

export interface Account {
    id: number;
    /**
     * The amount of money available in the account represented as a whole number using
     * the smallest monetary unit
     */
    balance: number;
    currency: Currency;
}
