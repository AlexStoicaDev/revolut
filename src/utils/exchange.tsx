export const getExchangedValue = (value: number, exchangeRate?: number): number => {
    return value * (exchangeRate || 1);
};
