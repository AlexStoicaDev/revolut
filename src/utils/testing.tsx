import React, { ReactElement, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import customTheme from '../theme';

export function renderWithTheme(component: ReactElement, options: any = {}) {
    const Wrapper: React.FC<{ children: ReactNode }> = ({ children }) => (
        <ThemeProvider theme={customTheme}>{children}</ThemeProvider>
    );

    return render(component, { wrapper: Wrapper, ...options });
}
