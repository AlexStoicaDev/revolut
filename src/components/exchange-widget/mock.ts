import { Currency } from '../../models/Currency';

interface Map {
    [key: string]: number | undefined;
}

const currencyPairs: Map = {
    GBPUSD: 1.38,
    GBPEUR: 1.17,
    EURUSD: 1.18,
    EURGBP: 0.83,
    USDGBP: 0.62,
    USDEUR: 0.82,
};

export const getExchangeRate = (currencyFrom: Currency, currencyTo: Currency) => {
    const currencyCode = currencyTo.code + currencyFrom.code;
    return currencyPairs[currencyCode];
};
export const currencyExchange = (amount: number, currencyTo: Currency, currencyFrom: Currency) => {
    const currencyCode = currencyTo.code + currencyFrom.code;
    return amount * (currencyPairs[currencyCode] || 1);
};
