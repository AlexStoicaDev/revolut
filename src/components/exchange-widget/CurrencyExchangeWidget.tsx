import { useEffect, useState } from 'react';
import { Account } from '../../models/Account';
import { currencyExchange, getExchangeRate } from './mock';
import Button from '../shared/button/Button';
import CurrencyExchangeInput from '../shared/currency-exchange-input/CurrencyExchangeInput';
import {
    CashflowArrow,
    ExchangeDescription,
    ExchangeRate,
    StyledCurrencyExchangeWidget,
} from './CurrencyExchangeWidget.styled';
import { AiOutlineStock } from 'react-icons/ai';

enum ExchangeAction {
    SELL = 'sell',
    BUY = 'buy',
}
const defaultInputState = {
    value: '',
    valueAsNumber: 0,
    exceedsBalance: false,
};

interface CurrencyExchangeWidgetProps {
    accounts: Account[];
}
const CurrencyExchangeWidget: React.FC<CurrencyExchangeWidgetProps> = ({ accounts }) => {
    const [firstSelectedAccount, setFirstSelectedAccount] = useState(accounts[0]);
    const [secondSelectedAccount, setSecondSelectedAccount] = useState(accounts[1]);
    const [firstCurrencyInputState, setFirstCurrencyInputState] = useState(defaultInputState);
    const [secondCurrencyInputState, setSecondCurrencyInputState] = useState(defaultInputState);
    const [exchangeAction, setExchangeAction] = useState<ExchangeAction>(ExchangeAction.SELL);
    const [exchangeRate, setExchangeRate] = useState(
        getExchangeRate(firstSelectedAccount.currency, secondSelectedAccount.currency),
    );

    useEffect(() => {
        setExchangeRate(
            getExchangeRate(firstSelectedAccount.currency, secondSelectedAccount.currency),
        );
    }, [firstSelectedAccount, secondSelectedAccount]);

    const firstInputExceedsBalance = (
        exchangeAction: ExchangeAction,
        valueToSubstract: number,
    ): boolean => {
        return (
            exchangeAction === ExchangeAction.SELL &&
            firstSelectedAccount.balance - valueToSubstract < 0
        );
    };
    const secondInputExceedsBalance = (
        exchangeAction: ExchangeAction,
        valueToSubstract: number,
    ): boolean => {
        return (
            exchangeAction === ExchangeAction.BUY &&
            secondSelectedAccount.balance - valueToSubstract < 0
        );
    };

    const onSelectAccount = (
        newAccountId: number,
        previousAccountId: number,
        theOtherAccountId: number,
        setterFunction: (account: Account) => void,
    ) => {
        if (newAccountId === previousAccountId) {
            return;
        }
        if (newAccountId === theOtherAccountId) {
            const aux = firstSelectedAccount;
            setFirstSelectedAccount(secondSelectedAccount);
            setSecondSelectedAccount(aux);
        } else {
            const newSelectedAccount = accounts.find(({ id }) => id === newAccountId);
            newSelectedAccount && setterFunction(newSelectedAccount);
        }
        setFirstCurrencyInputState(defaultInputState);
        setSecondCurrencyInputState(defaultInputState);
    };

    const onCashflowArrowClick = (newExchangeAction: ExchangeAction) => {
        setExchangeAction(newExchangeAction);
        setFirstCurrencyInputState((inputState) => ({
            ...inputState,
            exceedsBalance: firstInputExceedsBalance(newExchangeAction, inputState.valueAsNumber),
        }));
        setSecondCurrencyInputState((inputState) => ({
            ...inputState,
            exceedsBalance: secondInputExceedsBalance(newExchangeAction, inputState.valueAsNumber),
        }));
    };

    const handleFirstCurrencyInputChange = (value: string) => {
        const exhangeValue = currencyExchange(
            Number(value),
            firstSelectedAccount.currency,
            secondSelectedAccount.currency,
        );
        setFirstCurrencyInputState({
            value: value,
            valueAsNumber: Number(value),
            exceedsBalance: firstInputExceedsBalance(exchangeAction, Number(value)),
        });
        setSecondCurrencyInputState({
            value: exhangeValue.toString(),
            valueAsNumber: exhangeValue,
            exceedsBalance: secondInputExceedsBalance(exchangeAction, exhangeValue),
        });
    };

    const handleSecondCurrencyInputChange = (value: string) => {
        const exhangeValue = currencyExchange(
            Number(value),
            secondSelectedAccount.currency,
            firstSelectedAccount.currency,
        );
        setSecondCurrencyInputState({
            value: value,
            valueAsNumber: Number(value),
            exceedsBalance: secondInputExceedsBalance(exchangeAction, Number(value)),
        });
        setFirstCurrencyInputState({
            value: exhangeValue.toString(),
            valueAsNumber: exhangeValue,
            exceedsBalance: firstInputExceedsBalance(exchangeAction, exhangeValue),
        });
    };

    const getButtonText =
        exchangeAction === ExchangeAction.SELL
            ? `Sell ${firstSelectedAccount.currency.code} for ${secondSelectedAccount.currency.code}`
            : `Buy ${firstSelectedAccount.currency.code} with ${secondSelectedAccount.currency.code}`;

    const isButtonDisabled =
        firstCurrencyInputState.exceedsBalance ||
        secondCurrencyInputState.exceedsBalance ||
        firstCurrencyInputState.valueAsNumber === 0 ||
        secondCurrencyInputState.valueAsNumber === 0;

    return (
        <StyledCurrencyExchangeWidget>
            <ExchangeDescription data-testid="exchange-description">
                <h3>
                    {exchangeAction === ExchangeAction.SELL ? 'Sell ' : 'Buy '}
                    {firstSelectedAccount.currency.code}
                </h3>
            </ExchangeDescription>
            <ExchangeRate>
                <AiOutlineStock />
                <span data-testid="exchange-rate">{`${firstSelectedAccount.currency.symbol}1 = ${exchangeRate}${secondSelectedAccount.currency.symbol}`}</span>
            </ExchangeRate>
            <CurrencyExchangeInput
                availableAccounts={accounts}
                selectedAccount={firstSelectedAccount}
                onSelectAccount={(accountId: number) =>
                    onSelectAccount(
                        accountId,
                        firstSelectedAccount.id,
                        secondSelectedAccount.id,
                        setFirstSelectedAccount,
                    )
                }
                currencyInputValue={firstCurrencyInputState.value}
                exceedsBalance={firstCurrencyInputState.exceedsBalance}
                onCurrencyInputChange={handleFirstCurrencyInputChange}
            />
            <CashflowArrow
                data-testid="cashflow-arrow"
                $facingDownwards={!!(exchangeAction === ExchangeAction.SELL)}
                onClick={() =>
                    onCashflowArrowClick(
                        exchangeAction === ExchangeAction.SELL
                            ? ExchangeAction.BUY
                            : ExchangeAction.SELL,
                    )
                }
            />
            <CurrencyExchangeInput
                availableAccounts={accounts}
                selectedAccount={secondSelectedAccount}
                onSelectAccount={(accountId: number) =>
                    onSelectAccount(
                        accountId,
                        secondSelectedAccount.id,
                        firstSelectedAccount.id,
                        setSecondSelectedAccount,
                    )
                }
                currencyInputValue={secondCurrencyInputState.value}
                exceedsBalance={secondCurrencyInputState.exceedsBalance}
                onCurrencyInputChange={handleSecondCurrencyInputChange}
            />
            <Button type="submit" data-testid="exchange-button" disabled={isButtonDisabled}>
                <span>{getButtonText}</span>
            </Button>
        </StyledCurrencyExchangeWidget>
    );
};

export default CurrencyExchangeWidget;
