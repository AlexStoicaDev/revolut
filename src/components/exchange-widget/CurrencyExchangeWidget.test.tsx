import { cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Account } from '../../models/Account';
import { renderWithTheme } from '../../utils/testing';
import CurrencyExchangeWidget from './CurrencyExchangeWidget';
import { currencyExchange, getExchangeRate } from './mock';

afterEach(cleanup);

const currencies = [
    { code: 'GBP', symbol: '£' },
    { code: 'EUR', symbol: '€' },
    { code: 'USD', symbol: '$' },
];
const accounts: Account[] = [
    { id: 1, balance: 300, currency: currencies[0] },
    { id: 2, balance: 200, currency: currencies[1] },
    { id: 3, balance: 500, currency: currencies[2] },
];

const props = {
    accounts,
};

describe('CurrencyExchangeWidget', () => {
    it('should render the CurrencyExchangeWidget component correcly', () => {
        const { container } = renderWithTheme(<CurrencyExchangeWidget {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('can switch from sell to buy and vice versa', () => {
        const { getByTestId } = renderWithTheme(<CurrencyExchangeWidget {...props} />);
        const cashflowArrow = getByTestId('cashflow-arrow');
        const exchangeDescription = getByTestId('exchange-description');
        const exchangeRate = getByTestId('exchange-rate');
        const exchangeButton = getByTestId('exchange-button');
        const exchangeRateSnapshot = `
        <span
          data-testid="exchange-rate"
        >
          ${accounts[0].currency.symbol}1 = ${getExchangeRate(
            accounts[0].currency,
            accounts[1].currency,
        )}${accounts[1].currency.symbol}
        </span>
    `;
        expect(exchangeDescription.textContent).toMatchInlineSnapshot(
            `"Sell ${accounts[0].currency.code}"`,
        );
        expect(exchangeRate).toMatchInlineSnapshot(exchangeRateSnapshot);
        expect(exchangeButton.textContent).toMatchInlineSnapshot(
            `"Sell ${accounts[0].currency.code} for ${accounts[1].currency.code}"`,
        );

        userEvent.click(cashflowArrow);

        expect(exchangeDescription.textContent).toMatchInlineSnapshot(`"Buy GBP"`);
        expect(exchangeRate).toMatchInlineSnapshot(exchangeRateSnapshot);
        expect(exchangeButton.textContent).toMatchInlineSnapshot(
            `"Buy ${accounts[0].currency.code} with ${accounts[1].currency.code}"`,
        );

        userEvent.click(cashflowArrow);

        expect(exchangeDescription.textContent).toMatchInlineSnapshot(`"Sell GBP"`);
        expect(exchangeRate).toMatchInlineSnapshot(exchangeRateSnapshot);
        expect(exchangeButton.textContent).toMatchInlineSnapshot(
            `"Sell ${accounts[0].currency.code} for ${accounts[1].currency.code}"`,
        );
    });

    it('change in one of the inputs triggers convertion and change in the other', () => {
        const { getAllByTestId } = renderWithTheme(<CurrencyExchangeWidget {...props} />);
        const [firstCurrencyExchangeInput, secondCurrencyExchangeInput] =
            getAllByTestId('currency-input');

        userEvent.type(firstCurrencyExchangeInput, '2');
        userEvent.type(firstCurrencyExchangeInput, '1');

        // @ts-ignore
        expect(secondCurrencyExchangeInput.value).toBe(
            currencyExchange(21, accounts[0].currency, accounts[1].currency).toString(),
        );

        userEvent.clear(firstCurrencyExchangeInput);
        // @ts-ignore
        expect(secondCurrencyExchangeInput.value).toBe('0');
    });

    it('button id disabled when inputs are empty or one of them exceeds balance', () => {
        const { getByTestId, getAllByTestId } = renderWithTheme(
            <CurrencyExchangeWidget {...props} />,
        );
        const exchangeButton = getByTestId('exchange-button');
        const [firstCurrencyExchangeInput, secondCurrencyExchangeInput] =
            getAllByTestId('currency-input');
        const cashflowArrow = getByTestId('cashflow-arrow');

        // button is disabled when the inputs are empty
        expect(exchangeButton).toBeDisabled();

        // button is disabled when the first input exceeds balance
        userEvent.type(firstCurrencyExchangeInput, '1234');
        expect(exchangeButton).toBeDisabled();

        // button is disabled when the second input exceeds balance
        userEvent.click(cashflowArrow);
        userEvent.clear(firstCurrencyExchangeInput);
        userEvent.type(secondCurrencyExchangeInput, '1234');
        expect(exchangeButton).toBeDisabled();
    });

    it('button is enabled when the inputs contain correct values', () => {
        const { getByTestId, getAllByTestId } = renderWithTheme(
            <CurrencyExchangeWidget {...props} />,
        );
        const exchangeButton = getByTestId('exchange-button');
        const [firstCurrencyExchangeInput, secondCurrencyExchangeInput] =
            getAllByTestId('currency-input');
        expect(exchangeButton).toBeDisabled();

        userEvent.type(firstCurrencyExchangeInput, '12');
        userEvent.type(secondCurrencyExchangeInput, '34');
        expect(exchangeButton).not.toBeDisabled();
    });
});
