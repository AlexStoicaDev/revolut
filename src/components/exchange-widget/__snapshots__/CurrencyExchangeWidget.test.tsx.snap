// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyExchangeWidget should render the CurrencyExchangeWidget component correcly 1`] = `
<div>
  <form
    class="sc-kEjbQP cqa-DPr"
  >
    <div
      class="sc-crrszt XjWjY"
      data-testid="exchange-description"
    >
      <h3>
        Sell 
        GBP
      </h3>
    </div>
    <div
      class="sc-dQoVA inrupo"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M904 747H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM165.7 621.8l39.7 39.5c3.1 3.1 8.2 3.1 11.3 0l234.7-233.9 97.6 97.3a32.11 32.11 0 0 0 45.2 0l264.2-263.2c3.1-3.1 3.1-8.2 0-11.3l-39.7-39.6a8.03 8.03 0 0 0-11.3 0l-235.7 235-97.7-97.3a32.11 32.11 0 0 0-45.2 0L165.7 610.5a7.94 7.94 0 0 0 0 11.3z"
        />
      </svg>
      <span
        data-testid="exchange-rate"
      >
        £1 = 0.83€
      </span>
    </div>
    <div
      class="sc-gKseQn hYSsdt"
    >
      <div
        class="sc-iBPTik dJlHmu"
      >
        <div
          class="sc-gsTEea ifvwXS"
        >
          <div
            class="sc-dlfnuX itEhnx"
            data-testid="dropdown-handler"
          >
            <span
              data-testid="selected-currency"
            >
               
              GBP
               
            </span>
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 13.172l4.95-4.95 1.414 1.414L12 16 5.636 9.636 7.05 8.222z"
                />
              </g>
            </svg>
          </div>
        </div>
        <input
          class="sc-fubCzh hdFaqs"
          data-testid="currency-input"
          min="0"
          placeholder="0"
          step="1"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-pGacB gTpwTn"
      >
        <span
          data-testid="account-balance"
        >
          Balance: £ 300
        </span>
        <span
          class="sc-jrAFXE laYvyI"
          data-testid="balance-warning"
        >
          exceeds balance
        </span>
      </div>
    </div>
    <svg
      class="sc-iqHYmW iXiHPo"
      data-testid="cashflow-arrow"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.001,1.993C6.486,1.994,2,6.48,2,11.994c0.001,5.514,4.487,10,10,10c5.515,0,10.001-4.486,10.001-10 S17.515,1.994,12.001,1.993z M12,19.994c-4.41,0-7.999-3.589-8-8c0-4.411,3.589-8,8.001-8.001c4.411,0.001,8,3.59,8,8.001 S16.412,19.994,12,19.994z"
      />
      <path
        d="M12.001 8.001L7.996 12.006 11.001 12.006 11.001 16 13.001 16 13.001 12.006 16.005 12.006z"
      />
    </svg>
    <div
      class="sc-gKseQn hYSsdt"
    >
      <div
        class="sc-iBPTik dJlHmu"
      >
        <div
          class="sc-gsTEea ifvwXS"
        >
          <div
            class="sc-dlfnuX itEhnx"
            data-testid="dropdown-handler"
          >
            <span
              data-testid="selected-currency"
            >
               
              EUR
               
            </span>
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 13.172l4.95-4.95 1.414 1.414L12 16 5.636 9.636 7.05 8.222z"
                />
              </g>
            </svg>
          </div>
        </div>
        <input
          class="sc-fubCzh hdFaqs"
          data-testid="currency-input"
          min="0"
          placeholder="0"
          step="1"
          type="number"
          value=""
        />
      </div>
      <div
        class="sc-pGacB gTpwTn"
      >
        <span
          data-testid="account-balance"
        >
          Balance: € 200
        </span>
        <span
          class="sc-jrAFXE laYvyI"
          data-testid="balance-warning"
        >
          exceeds balance
        </span>
      </div>
    </div>
    <button
      class="sc-bdfBQB jjMWNz"
      data-testid="exchange-button"
      disabled=""
      type="submit"
    >
      <span>
        Sell GBP for EUR
      </span>
    </button>
  </form>
</div>
`;
